design/one-pager-merge-jets-proposal.md

```diff
@@ -0,0 +1,432 @@
+knows which field should be used as the source of information. Take a look at
+the following example:
+
+```yaml
+# User can give this information manually but if vpcIdRef is used, controller knows
+# which field of VPC object to copy here.
+vpcId: vpc-2314334
+# User can give only the name of the VPC object and then vpcId will be populated.
+vpcIdRef:
+  name: my-vpc
+# User can give only a selector and then the vpcIdRef will be populated.
+vpcIdSelector:
+  matchLabels:
+    app: ola
+```
+
+When we have Jet and classic providers, one does not know about the other's
+types. Hence, the only way to do references is to give the type of the
+referenced objct and the field path of the information on that type, something
+roughly like the following:
+
+```yaml
+vpcId: vpc-2314334
+vpcIdRef:
+  apiVersion: ec2.aws.crossplane.io/v1beta1
+  kind: VPC
+  fieldPath: metadata.annotations[crossplane.io/external-name]
+vpcIdSelector:
+  apiVersion: ec2.aws.crossplane.io/v1beta1
+  kind: VPC
+  fieldPath: metadata.annotations[crossplane.io/external-name]
+  matchLabels:
+    app: ola
+```
+
+While this referencing style is useful for many scenarios and being discussed
+[here][generic-references], it's not the same as how Kubernetes resources
+reference each other and requires more upfront information. We could possibly
+make it work like native with a default and let users optionally give different
+`kind`, `apiVersion` and `fieldPath`, there are other concerns with the approach
+listed in its discussion.
+
+In a single provider, users would keep using Kubernetes-native style referencing
+where they are only concerned with the name and/or labels of the referenced
+object, since there is a single type and it's defined in the same provider, the
+controller knows these details already.
+
+## Drawbacks
+
+### Code Difference
+
+Terrajet has a single controller implementation that you configure to work with
+different CRDs but the folder structure is the same as classic providers, i.e.
+`apis` and `pkg/controller` folders and the rest is exactly the same. The most
+apparent difference is that in classic providers, you have the controller
+`Setup` function [together with the implementation][classic-setup] of the
+controller whereas in Jet providers, there is only `Setup` [function][jet-setup]
+since its implementation lives in `terrajet` repository. In either case, the
+controller implementation satisfies the same
+[`ExternalClient`][externalclient-interface] interface from crossplane-runtime.
+
+The biggest difference in that regard is that the CRD structs in Jet providers
+need to satisfy an additional interface called
+[`Terraformed`][terraformed-interface] in order for the generic controller to
+work. So, even though the structure is same, you'll see that [the generic
+controller][generic-controller] makes extensive use of the functions defined in
+that interface where as classic implementations work with the type directly.
```

Comment:
I couldn't get why this is a drawback, given that most Terrajet code is auto-generated. I can understand that maintainers should understand how Terrajet works due to this difference but it looks like this is already covered below.